### DISCUSSION TITLE
Prisma Client not working properly in Docker Container

### DISCUSSION THREAD

**Discussion by Story27** [prisma/prisma#24885] [Q&A] [answered]
### Question

this is how i am generating prisma client

`import { PrismaClient } from "@prisma/client";
export const db = new PrismaClient();
`

This is my project Directory
```
project-root/
├── app/
│   ├── components/
│   ├── pages/
│   ├── styles/
│   └── ... (other app-specific directories and files)
├── prisma/
│   ├── migrations/
│   ├── schema.prisma
│   └── ... (other Prisma-specific files)
├── node_modules/
├── .next/
├── .vscode/
├── Dockerfile
├── .dockerignore
├── README.md
├── npm-debug.log
├── .git/
├── instructions.txt
├── package.json
├── tsconfig.json (if using TypeScript)
├── next.config.js (if using Next.js)
└── ... (other project-specific files)
```

Everything in Development is working as I want to but when I am converting this to a docker container all POST request are getting saved in the database (I am using PostgreSQL on Neon), but they are not shown when I'm accessing those data from api endpoints on certain pages. Like it is working for this page

```
useEffect(() => {
    const fetchProblem = async () => {
      try {
        const response = await fetch(`/api/problems/${problemId}`);
        if (!response.ok) throw new Error("Failed to fetch problem data");
        const data = await response.json();
        setProblem(data);
      } catch (err) {
        console.error("Error fetching problem data:", err);
        setError("Failed to load problem data. Please try again.");
      }
    };

    if (problemId) {
      fetchProblem();
    }
  }, [problemId]);

```

but not working on this page

```
React.useEffect(() => {
    async function fetchData() {
      const response = await fetch("/api/problems");
      const result = await response.json();
      setData(result);
      setLoading(false);
    }
    fetchData();
  }, []);
```

Its working like all things that were there initially in the database when the docker container was made, it stays like no edits are occurring on the database and I think thats because of wrong Prisma Client generation.

This is my dockerfile

```
FROM node:20-alpine AS base

# Install dependencies only when needed
FROM base AS deps
RUN apk add --no-cache libc6-compat g++ python3 openjdk11
WORKDIR /app

# Install dependencies
COPY package.json package-lock.json* ./
RUN npm ci

# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY --from=deps /app/package-lock.json ./package-lock.json
COPY . .
COPY prisma ./prisma

ENV NEXT_TELEMETRY_DISABLED 1

# Generate Prisma client
RUN npx prisma generate --schema=./prisma/schema.prisma
RUN npx prisma db push
RUN npx prisma migrate deploy

RUN npm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

# Install g++ in the runner stage
RUN apk add --no-cache g++

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Create necessary directories and set permissions
RUN mkdir -p /app/Executed-codes/codes /app/Executed-codes/outputs /app/Executed-codes/inputs
RUN chown -R nextjs:nodejs /app/Executed-codes

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

# Copy Prisma files
COPY --from=builder /app/prisma ./prisma
COPY --from=builder /app/node_modules/.prisma ./node_modules/.prisma
COPY --from=builder /app/node_modules/@prisma ./node_modules/@prisma

# Generate Prisma client again in the runner stage
RUN npx prisma generate --schema=prisma/schema.prisma
RUN npx prisma db push
RUN npx prisma migrate deploy

USER nextjs

EXPOSE 3000

ENV PORT 3000
ENV HOSTNAME "0.0.0.0"

# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
```

I am linking my GitHub Repo so that you'll can check the rest of the code if any other problem is there
[GitHub Repo](https://github.com/Story27/Online-Judge)

### Expected behavior (optional)

I want that inside docker container all GET requests on api endpoints have updated data as updated in the database.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
datasource db {
  provider = "postgresql"
  url = env("POSTGRES_PRISMA_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id                    String                 @id @default(cuid())
  name                  String?
  email                 String?                @unique
  emailVerified         DateTime?
  image                 String?
  password              String?
  role                  UserRole               @default(USER)
  accounts              Account[]
  isTwoFactorEnabled    Boolean                @default(false)
  twoFactorConfirmation TwoFactorConfirmation?
  problemSets           Problem[]

  Contest Contest[]
}

model Account {
  id                String  @id @default(cuid())
  userId            String
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.Text
  access_token      String? @db.Text
  expires_at        Int?
  token_type        String?
  scope             String?
  id_token          String?
  session_state     String?

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model VerificationToken {
  id      String   @id @default(cuid())
  email   String
  token   String   @unique
  expires DateTime

  @@unique([email, token])
}

model PasswordResetToken {
  id      String   @id @default(cuid())
  email   String
  token   String   @unique
  expires DateTime

  @@unique([email, token])
}

model TwoFactorToken {
  id      String   @id @default(cuid())
  email   String
  token   String   @unique
  expires DateTime

  @@unique([email, token])
}

model TwoFactorConfirmation {
  id     String @id @default(cuid())
  userId String
  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([userId])
}

model Contest {
  id        String    @id @default(cuid())
  name      String
  startTime DateTime
  endTime   DateTime
  userId    String
  user      User      @relation(fields: [userId], references: [id], onDelete: Cascade)
  problems  Problem[]
}

model Problem {
  id          String     @id @default(cuid())
  title       String
  description String?    @default("")
  difficulty  Difficulty
  topics      String
  testCases   TestCase[]
  userId      String
  user        User       @relation(fields: [userId], references: [id], onDelete: Cascade)
  contests    Contest[]
  acceptances String[] @default([])
}

model TestCase {
  id               String  @id @default(cuid())
  input            String
  output           String
  problemId        String
  problem          Problem @relation(fields: [problemId], references: [id])
  isSampleTestCase Boolean @default(false)
}

enum UserRole {
  ADMIN
  USER
}

enum Difficulty {
  EASY
  MEDIUM
  HARD
}

```

this is one of the POST query
```ts
"use server";

import { db } from "@/lib/db";
import { ProblemSchema, ProblemInput } from "@/schemas";
import { UserRole } from "@prisma/client";

export const addProblem = async (
  data: ProblemInput,
  userId: string,
  userRole: UserRole
) => {
  console.log("Received data:", data);
  const validatedFields = ProblemSchema.safeParse({ ...data, userId });

  if (!validatedFields.success) {
    console.error("Validation Error:", validatedFields.error);
    return { error: "Invalid Fields!" };
  }

  const { title, difficulty, topics, description, testCases } =
    validatedFields.data;

  // Check user role
  if (userRole !== UserRole.ADMIN) {
    return { error: "Forbidden" };
  }

  try {
    // Create problem with basic details
    const createdProblem = await db.problem.create({
      data: {
        title,
        difficulty,
        topics,
        description,
        userId,
        testCases: {
          create: testCases.map((testCase) => ({
            input: testCase.input,
            output: testCase.output,
            isSampleTestCase: testCase.isSampleTestCase,
          })),
        },
      },
      include: {
        testCases: true,
      },
    });

    return { success: "Problem created successfully!" };
  } catch (error) {
    console.error("Error creating problem:", error);
    return { error: "Failed to create problem." };
  }
};
```

- OS: Windows
- Database: PostgreSQL
- Node.js version: v22.5.0

- Run `prisma -v` to see your Prisma version and paste it
Environment variables loaded from .env
prisma                  : 5.16.2
@prisma/client          : 5.16.2
Computed binaryTarget   : windows
Operating System        : win32
Architecture            : x64
Node.js                 : v22.5.0
Query Engine (Node-API) : libquery-engine 34ace0eb2704183d2c05b60b52fba5c43c13f303 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 34ace0eb2704183d2c05b60b52fba5c43c13f303 (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.16.0-24.34ace0eb2704183d2c05b60b52fba5c43c13f303
Default Engines Hash    : 34ace0eb2704183d2c05b60b52fba5c43c13f303
Studio                  : 0.502.0

Here is the image how the data is shown updated locally
![image](https://github.com/user-attachments/assets/33d5f1c3-030c-40f5-bb4a-ec3ca7bade2e)

and this is how it is shown on docker container it is showing the values that were there when the container was made nothing has been updating on it
![image](https://github.com/user-attachments/assets/bfa8ef6e-bdd2-4ca8-9b59-23bed76c9f34)

**Comment by ludralph** [comment, marked as answer]
Hi @Story27 👋 

In your docker file, you are generating the Prisma client multiple times. This is unnecessary and can be optimized. You should generate the Prisma client only once during the build stage. You should also ensure that the `node_modules` directory is correctly copied from the builder stage to the runner stage. Try out this revised docker file:
```
FROM node:20-alpine AS base

# Install dependencies only when needed
FROM base AS deps
RUN apk add --no-cache libc6-compat g++ python3 openjdk11
WORKDIR /app

# Install dependencies
COPY package.json package-lock.json* ./
RUN npm ci

# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY --from=deps /app/package-lock.json ./package-lock.json
COPY . .
COPY prisma ./prisma

ENV NEXT_TELEMETRY_DISABLED 1

# Generate Prisma client
RUN npx prisma generate --schema=./prisma/schema.prisma

RUN npm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

# Install g++ in the runner stage
RUN apk add --no-cache g++

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Create necessary directories and set permissions
RUN mkdir -p /app/Executed-codes/codes /app/Executed-codes/outputs /app/Executed-codes/inputs
RUN chown -R nextjs:nodejs /app/Executed-codes

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

# Copy Prisma files
COPY --from=builder /app/prisma ./prisma
COPY --from=builder /app/node_modules/.prisma ./node_modules/.prisma
COPY --from=builder /app/node_modules/@prisma ./node_modules/@prisma

USER nextjs

EXPOSE 3000

ENV PORT 3000
ENV HOSTNAME "0.0.0.0"

# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
```

**Comment by Story27** [comment]
Thanks @ludralph  for your help this code worked.

**Reply by ludralph** [reply]
I'm glad i could help 🙏

**Comment by adityadeshlahre** [comment]
Hello @Story27 @ludralph I am finding a bit difficult to use prisma client insider docker container its getting generated properly but when ever `prisma generate` command is run inside docker it is breaking! 

some context : I am trying to build my own starter kit with next js which can be deployed in anywhere as final distribution will happened through docker my repo [next-start](https://github.com/adityadeshlahre/next-start)

my docker file 

```
FROM node:22-alpine AS base

FROM base AS builder
ENV PNPM_HOME="/root/.local/share/pnpm"
ENV PATH="${PATH}:${PNPM_HOME}"

WORKDIR /app
RUN npm install --global pnpm@latest
RUN pnpm add -g turbo
COPY . .
RUN turbo prune --scope=web --docker

FROM base AS installer

ENV PNPM_HOME="/root/.local/share/pnpm"
ENV PATH="${PATH}:${PNPM_HOME}"

RUN npm install --global pnpm@latest

WORKDIR /app

COPY .gitignore .gitignore
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/pnpm-lock.yaml ./pnpm-lock.yaml
COPY --from=builder /app/out/full/ .
COPY turbo.json turbo.json

RUN pnpm i

#PRISMA
RUN pnpx prisma generate --schema=./packages/database/prisma/schema.prisma
RUN mkdir -p /app/.prisma
RUN cp -r ./node_modules/.pnpm/@prisma+client@6.0.0_prisma@6.0.0/node_modules/@prisma/client /app/.prisma
RUN pnpm turbo run build --filter=web...

FROM base AS runner

ENV PNPM_HOME="/root/.local/share/pnpm"
ENV PATH="${PATH}:${PNPM_HOME}"

RUN npm install --global pnpm@latest

WORKDIR /app

# COPY --from=installer /app .
COPY --from=installer /app/apps/web/.next/standalone ./
COPY --from=installer /app/apps/web/.next/static ./apps/web/.next/static
# COPY --from=installer /app/apps/web/.next/static ./apps/web/.next/standalone/apps/web/.next/static
COPY --from=installer /app/apps/web/public ./apps/web/public
# COPY --from=installer /app/apps/web/public ./apps/web/.next/standalone/apps/web/public


RUN mkdir -p /app/node_modules/@prisma

#PRISMA
COPY --from=installer /app/.prisma /app/.prisma
COPY --from=installer /app/.prisma /app/node_modules/.prisma
# COPY --from=installer /app/.prisma /app/node_modules/@prisma

ENV PORT=3000
EXPOSE 3000
ENV HOSTNAME=0.0.0.0

# WORKDIR /app/apps/web

# CMD ["pnpm", "dev"]
# CMD node apps/web/server.js
# CMD  HOSTNAME="0.0.0.0" node ./apps/web/server.js
CMD ["node", "./apps/web/.next/standalone/apps/web/server.js"]
```

I am very much confused trying things to make it work and hence reached here to have professional answers 

ERROR : image 

![image](https://github.com/user-attachments/assets/8679ed56-e8f4-4d81-ae19-d36fffd7e9f8)

This is the error I face while running the application from inside of the docker

**Reply by luizgustavosaraiva** [reply]
The

> Hello @Story27 @ludralph I am finding a bit difficult to use prisma client insider docker container its getting generated properly but when ever `prisma generate` command is run inside docker it is breaking!
> 
> some context : I am trying to build my own starter kit with next js which can be deployed in anywhere as final distribution will happened through docker my repo [next-start](https://github.com/adityadeshlahre/next-start)
> 
> my docker file
> 
> ```
> FROM node:22-alpine AS base
> 
> FROM base AS builder
> ENV PNPM_HOME="/root/.local/share/pnpm"
> ENV PATH="${PATH}:${PNPM_HOME}"
> 
> WORKDIR /app
> RUN npm install --global pnpm@latest
> RUN pnpm add -g turbo
> COPY . .
> RUN turbo prune --scope=web --docker
> 
> FROM base AS installer
> 
> ENV PNPM_HOME="/root/.local/share/pnpm"
> ENV PATH="${PATH}:${PNPM_HOME}"
> 
> RUN npm install --global pnpm@latest
> 
> WORKDIR /app
> 
> COPY .gitignore .gitignore
> COPY --from=builder /app/out/json/ .
> COPY --from=builder /app/out/pnpm-lock.yaml ./pnpm-lock.yaml
> COPY --from=builder /app/out/full/ .
> COPY turbo.json turbo.json
> 
> RUN pnpm i
> 
> #PRISMA
> RUN pnpx prisma generate --schema=./packages/database/prisma/schema.prisma
> RUN mkdir -p /app/.prisma
> RUN cp -r ./node_modules/.pnpm/@prisma+client@6.0.0_prisma@6.0.0/node_modules/@prisma/client /app/.prisma
> RUN pnpm turbo run build --filter=web...
> 
> FROM base AS runner
> 
> ENV PNPM_HOME="/root/.local/share/pnpm"
> ENV PATH="${PATH}:${PNPM_HOME}"
> 
> RUN npm install --global pnpm@latest
> 
> WORKDIR /app
> 
> # COPY --from=installer /app .
> COPY --from=installer /app/apps/web/.next/standalone ./
> COPY --from=installer /app/apps/web/.next/static ./apps/web/.next/static
> # COPY --from=installer /app/apps/web/.next/static ./apps/web/.next/standalone/apps/web/.next/static
> COPY --from=installer /app/apps/web/public ./apps/web/public
> # COPY --from=installer /app/apps/web/public ./apps/web/.next/standalone/apps/web/public
> 
> 
> RUN mkdir -p /app/node_modules/@prisma
> 
> #PRISMA
> COPY --from=installer /app/.prisma /app/.prisma
> COPY --from=installer /app/.prisma /app/node_modules/.prisma
> # COPY --from=installer /app/.prisma /app/node_modules/@prisma
> 
> ENV PORT=3000
> EXPOSE 3000
> ENV HOSTNAME=0.0.0.0
> 
> # WORKDIR /app/apps/web
> 
> # CMD ["pnpm", "dev"]
> # CMD node apps/web/server.js
> # CMD  HOSTNAME="0.0.0.0" node ./apps/web/server.js
> CMD ["node", "./apps/web/.next/standalone/apps/web/server.js"]
> ```
> 
> I am very much confused trying things to make it work and hence reached here to have professional answers
> 
> ERROR : image
> 
> ![image](https://private-user-images.githubusercontent.com/132184385/391654012-8679ed56-e8f4-4d81-ae19-d36fffd7e9f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2NDU0NTAsIm5iZiI6MTczNzY0NTE1MCwicGF0aCI6Ii8xMzIxODQzODUvMzkxNjU0MDEyLTg2NzllZDU2LWU4ZjQtNGQ4MS1hZTE5LWQzNmZmZmQ3ZTlmOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyM1QxNTEyMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYTkwNzg3NTU1NTY3MjhhMzYzYTFmYWIzMzdmOTMxNjM4OGI1ODQ3OWQ5YWM0N2U1NzVkYTE5ZTAxOGI2ZjhhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ynoaBw-yRXckzbLosfwKUheIo9bogG6hU3SU3gP1yMU)
> 
> This is the error I face while running the application from inside of the docker

The cause of this binary error is the version of your Alpine; according to them, the OpenSSL binaries have changed the base path, and it has not been resolved yet. I recommend using the version `node:20-alpine3.16`

https://github.com/nodejs/docker-node/issues/2175
